pkg/apis/pipeline/v1/matrix_types.go

```diff
@@ -52,50 +53,45 @@ type MatrixInclude struct {
 }
 
 // FanOut produces combinations of Parameters of type String from a slice of Parameters of type Array.
-func (m *Matrix) FanOut() []Params {
-	var combinations []Params
+func (m *Matrix) FanOut() []map[string]string {
+	var combinations []map[string]string
 	for _, parameter := range m.Params {
 		combinations = fanOut(parameter, combinations)
 	}
 	return combinations
 }
 
 // fanOut generates a new combination based on a given Parameter in the Matrix.
-func fanOut(param Param, combinations []Params) []Params {
+func fanOut(param Param, combinations []map[string]string) []map[string]string {
 	if len(combinations) == 0 {
 		return initializeCombinations(param)
 	}
 	return distribute(param, combinations)
 }
 
 // distribute generates a new combination of Parameters by adding a new Parameter to an existing list of Combinations.
-func distribute(param Param, combinations []Params) []Params {
-	var expandedCombinations []Params
+func distribute(param Param, combinations []map[string]string) []map[string]string {
```

Comment:
wondering if we can define `[]map[string]string` as `combinations` so that we can create member functions on it, for example:

 ```suggestion
 type combinations []map[string]string
 
func (c *combinations) distribute(param Param) combinations {
```

Reply:
thanks @jerop I can look into it. I have been talking to @EmmaMunley to utilize the same `Matrix` struct instead of creating a new `struct` to hold all possible combinations for the given `matrix`. This new field in the existing struct will not be exposed to the users. Thoughts?

Reply:
sounds good 👍🏾

Reply:
I love to hear the significance of member functions 🎉 makes the code so much cleaner and maintainable 🎉